### PR TITLE
Ensure TripalFields can be added through UI and add some tests

### LINKED ISSUE #1966: Tripal (not chado) fields can't be added
### Tripal Version

4.x

### Drupal Version

10.2, 10.3

### PostgreSQL Version

13, 16

### PHP Version

8.3

### General Description

I can't add a tripal field 🙈

### Steps to reproduce

Tripal -> Page Structure -> Project -> Manage Fields
+Create a new field -> Tripal Fields -> Continue
Label enter anything
Check any of the four field types -> Continue
WSOD as shown below

### Error messages

The website encountered an unexpected error. Try again later.

Exception: Cannot create a StorageProperty object for entity type "tripal_entity", field type "tripal_integer_type", key "value" without a properly formatted term: ":" in Drupal\tripal\TripalStorage\StoragePropertyBase->__construct() (line 83 of modules/contrib/tripal/tripal/src/TripalStorage/StoragePropertyBase.php).

```
Drupal\tripal\TripalStorage\StoragePropertyTypeBase->__construct('tripal_entity', 'tripal_integer_type', 'value', ':', 'int', Array) (Line: 32)
Drupal\tripal\TripalStorage\IntStoragePropertyType->__construct('tripal_entity', 'tripal_integer_type', 'value', ':') (Line: 37)
Drupal\tripal\Plugin\Field\FieldType\TripalIntegerTypeItem::tripalTypes(Object) (Line: 401)
Drupal\tripal\TripalField\TripalFieldItemBase::propertyDefinitions(Object) (Line: 794)
Drupal\field\Entity\FieldStorageConfig->getPropertyDefinitions() (Line: 65)
Drupal\Core\Field\TypedData\FieldItemDataDefinition->getPropertyDefinitions() (Line: 64)
Drupal\Core\Field\FieldItemBase->__construct(Object, 0, Object) (Line: 47)
Drupal\Core\TypedData\TypedData::createInstance(Object, 0, Object) (Line: 99)
Drupal\Core\TypedData\TypedDataManager->createInstance('field_item:tripal_integer_type', Array) (Line: 108)
Drupal\Core\TypedData\TypedDataManager->create(Object, NULL, 0, Object) (Line: 208)
Drupal\Core\TypedData\TypedDataManager->getPropertyInstance(Object, 0, NULL) (Line: 100)
Drupal\Core\Field\FieldTypePluginManager->createFieldItem(Object, 0, NULL) (Line: 41)
Drupal\Core\Field\FieldItemList->createItem(0, NULL) (Line: 214)
Drupal\Core\TypedData\Plugin\DataType\ItemList->appendItem() (Line: 122)
Drupal\field_ui\Form\FieldStorageConfigEditForm->form(Array, Object) (Line: 107)
Drupal\Core\Entity\EntityForm->buildForm(Array, Object) (Line: 85)
Drupal\field_ui\Form\FieldStorageConfigEditForm->buildForm(Array, Object, Object) (Line: 199)
Drupal\field_ui\Form\FieldConfigEditForm->form(Array, Object) (Line: 107)
Drupal\Core\Entity\EntityForm->buildForm(Array, Object)
call_user_func_array(Array, Array) (Line: 536)
Drupal\Core\Form\FormBuilder->retrieveForm('field_config_edit_form', Object) (Line: 284)
Drupal\Core\Form\FormBuilder->buildForm(Object, Object) (Line: 48)
Drupal\Core\Entity\EntityFormBuilder->getForm(Object, 'default', Array) (Line: 62)
Drupal\field_ui\Controller\FieldConfigAddController->fieldConfigAddConfigureForm('tripal_entity', 'field_tf')
call_user_func_array(Array, Array) (Line: 123)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}() (Line: 638)
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object) (Line: 121)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext(Array, Array) (Line: 97)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}() (Line: 181)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1) (Line: 76)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1) (Line: 53)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1) (Line: 48)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1) (Line: 28)
Drupal\Core\StackMiddleware\ContentLength->handle(Object, 1, 1) (Line: 32)
Drupal\big_pipe\StackMiddleware\ContentLength->handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1) (Line: 48)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1) (Line: 51)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1) (Line: 36)
Drupal\Core\StackMiddleware\AjaxPageState->handle(Object, 1, 1) (Line: 51)
Drupal\Core\StackMiddleware\StackedHttpKernel->handle(Object, 1, 1) (Line: 741)
Drupal\Core\DrupalKernel->handle(Object) (Line: 19)
```

### PR DESCRIPTION
**Pull request by laceysanderson**
# Bug Fix

## Description

In the linked issue @dsenalik pointed out that TripalFields could not be added throguh the UI. I was already doing some automated tests for TripalFields in another branch so added the fix to that issue in with my testing to create this PR.

## Testing?

1. Tripal -> Page Structure -> Project -> Manage Fields
2. +Create a new field -> Tripal Fields -> Continue
3. Label enter anything
4. Check any of the four field types -> Continue
5. Confirm the new field is now listed in the fields for project.
6. Create a project with a number of values including a value for your new Tripal field instance and confirm it saves.
7. Edit this project, change a bunch of values including the value of your new TripalField and confirm it is updated.
8. Also create and edit content where the TripalField does not have a value entered.

### Automated Testing

This PR adds some basic testing for TripalFields. The tests in Drupal\Tests\tripal\Kernel\TripalField\TripalFieldTypeSettingsTest check that the subform for the field storage settings can be built. This test fails without the fix.

This PR also adds some basic testing of the Tripal field types in Drupal\Tests\tripal\Kernel\TripalField\TripalFieldTypeCRUDTest although it currently only tests create. Tests for edit and delete will be coming in another PR.

Screenshot 2024-09-03 at 2 17 06 PM
Screenshot 2024-09-09 at 10 34 57 AM

### CODE REVIEW

File: tripal/src/Plugin/Field/FieldType/TripalBooleanTypeItem.php [hunk 1]
```diff
@@ -24,6 +24,36 @@ class TripalBooleanTypeItem extends TripalFieldItemBase {
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function generateSampleValue(FieldDefinitionInterface $field_definition) {
+    $values = [];
+
+    $random = new \Drupal\Component\Utility\Random();
+    $values['value'] = TRUE;
```
Comment (dsenalik) [inline tripal/src/Plugin/Field/FieldType/TripalBooleanTypeItem.php]: Do we want to generate TRUE half the time and FALSE half the time here?

**Review by dsenalik — APPROVED** [review]
One comment from code review.
Manual testing worked as advertised. I did all four tripal fields. One comment I have, and I think this belongs in another issue maybe, the boolean field, if it is not set to TRUE, does not display on the entity page. We have a display setting (tested that too) to set what to display for each TRUE and FALSE. I feel setting the false value to an empty string should cause it to be hidden, but otherwise it should display 'false' or whatever.
issue-1966-works

Edit
The chado boolean field works like this
chadoboolean

Currently you can't set it to display an empty string
chadoboolean2

Your comment https://github.com/tripal/tripal/pull/1952#issuecomment-2327320716 so this could be a new issue.
"I think we should leave a default of 0 for integer fields and create a new issue for this. I think I would be tempted to solve it with an option in the formatter for integer fields that lets you indicate if you want to hide entries with a value of 0 or not 🤔 but there may be a better option."

File: tripal/src/Plugin/Field/FieldType/TripalBooleanTypeItem.php (hunk 1 above)
Comment (laceysanderson) [reply tripal/src/Plugin/Field/FieldType/TripalBooleanTypeItem.php]: I thought about this but must have forgotten before taking the time to actually implement it 🙈 This is now fixed in d064e4374.

**Comment by laceysanderson** [comment]
Thanks for the review @dsenalik ❤️ I agree that that should be a new issue 🙈

**Comment by dsenalik** [comment]
> Thanks for the review @dsenalik ❤️ I agree that that should be a new issue 🙈

This is now issue ~~#1969~~  #1970

**Pull request merged**